### ISSUE TITLE
3.0.0-20200106.122158-51 breaks - Rest Data

### ISSUE THREAD

**Issue by Syndlex** [springfox/springfox#3227]
Please take the time to search the repository, if your question has already been asked or answered.

- [X] What version of the library are you using? 
3.0.0 Snapshot

What kind of issue is this?
Setup with Spring Data is breaking in latest Snapshot
 - [X] Bug report. 
Spring has problems with the Hateos Link Provider.
    - [x] include any output you've received; logs, json snippets etc.; and what the expected output
```
java.lang.IllegalStateException: Failed to introspect Class [springfox.documentation.spring.data.rest.schema.SpringDataRestSchemaExtensions] from ClassLoader [jdk.internal.loader.ClassLoaders$AppClassLoader@77556fd]
	at org.springframework.util.ReflectionUtils.getDeclaredMethods(ReflectionUtils.java:507) ~[spring-core-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.util.ReflectionUtils.doWithMethods(ReflectionUtils.java:404) ~[spring-core-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.util.ReflectionUtils.doWithMethods(ReflectionUtils.java:389) ~[spring-core-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.util.ReflectionUtils.getUniqueDeclaredMethods(ReflectionUtils.java:447) ~[spring-core-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705) ~[na:na]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.getTypeForFactoryMethod(AbstractAutowireCapableBeanFactory.java:737) ~[spring-beans-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.determineTargetType(AbstractAutowireCapableBeanFactory.java:678) ~[spring-beans-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.predictBeanType(AbstractAutowireCapableBeanFactory.java:647) ~[spring-beans-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.isFactoryBean(AbstractBeanFactory.java:1518) ~[spring-beans-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doGetBeanNamesForType(DefaultListableBeanFactory.java:513) ~[spring-beans-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeanNamesForType(DefaultListableBeanFactory.java:483) ~[spring-beans-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeansOfType(DefaultListableBeanFactory.java:604) ~[spring-beans-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeansOfType(DefaultListableBeanFactory.java:596) ~[spring-beans-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.getBeansOfType(AbstractApplicationContext.java:1226) ~[spring-context-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	at org.springframework.boot.SpringApplication.getExitCodeFromMappedException(SpringApplication.java:866) ~[spring-boot-2.1.11.RELEASE.jar:2.1.11.RELEASE]
	at org.springframework.boot.SpringApplication.getExitCodeFromException(SpringApplication.java:854) ~[spring-boot-2.1.11.RELEASE.jar:2.1.11.RELEASE]
	at org.springframework.boot.SpringApplication.handleExitCode(SpringApplication.java:841) ~[spring-boot-2.1.11.RELEASE.jar:2.1.11.RELEASE]
	at org.springframework.boot.SpringApplication.handleRunFailure(SpringApplication.java:792) ~[spring-boot-2.1.11.RELEASE.jar:2.1.11.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:322) ~[spring-boot-2.1.11.RELEASE.jar:2.1.11.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1215) ~[spring-boot-2.1.11.RELEASE.jar:2.1.11.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1204) ~[spring-boot-2.1.11.RELEASE.jar:2.1.11.RELEASE]
	at de.hda.y.vier.cloudengpraktikum.CloudEngPraktikum.main(CloudEngPraktikum.java:12) ~[main/:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.springframework.boot.devtools.restart.RestartLauncher.run(RestartLauncher.java:49) ~[spring-boot-devtools-2.1.11.RELEASE.jar:2.1.11.RELEASE]
Caused by: java.lang.NoClassDefFoundError: org/springframework/hateoas/server/LinkRelationProvider
	at java.base/java.lang.Class.getDeclaredMethods0(Native Method) ~[na:na]
	at java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3166) ~[na:na]
	at java.base/java.lang.Class.getDeclaredMethods(Class.java:2309) ~[na:na]
	at org.springframework.util.ReflectionUtils.getDeclaredMethods(ReflectionUtils.java:489) ~[spring-core-5.1.12.RELEASE.jar:5.1.12.RELEASE]
	... 26 common frames omitted
Caused by: java.lang.ClassNotFoundException: org.springframework.hateoas.server.LinkRelationProvider
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581) ~[na:na]
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178) ~[na:na]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521) ~[na:na]
```
- [x] spring xml/java config that is relevant
     Spring Fox Config
```
@Configuration
@EnableSwagger2WebMvc
@Import({  springfox.documentation.spring.data.rest.configuration.SpringDataRestConfiguration.class})
public class SpringFox {

    /**
     * Swagger UI configuration.
     *
     * @return
     */
    @Bean
    //Visible under http://localhost:8080/swagger-ui.html
    public Docket apiDocket() {
        return new Docket(DocumentationType.SWAGGER_2)
                .select()
                .paths(Predicate.not(PathSelectors.regex("/error")))
                .paths(Predicate.not(PathSelectors.regex("/profile")))
                .paths(PathSelectors.any())
                .build()
                .useDefaultResponseMessages(false);
    }
}
```
Gradle Config Maybe Interesting

```
plugins {
    id 'org.springframework.boot' version '2.1.11.RELEASE'
    id 'io.spring.dependency-management' version '1.0.8.RELEASE'
    id 'java'
}

repositories {
    mavenCentral()
    maven { url 'http://oss.jfrog.org/artifactory/oss-snapshot-local/' }
}

group = 'de.XXX.y.XX'
version = '0.0.1-SNAPSHOT'
sourceCompatibility = JavaVersion.VERSION_11

configurations {
    developmentOnly
    runtimeClasspath {
        extendsFrom developmentOnly
    }
    compileOnly {
        extendsFrom annotationProcessor
    }
}
ext {
    set('springCloudVersion', "Greenwich.SR4")
}

dependencies {
    implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
    implementation 'org.springframework.boot:spring-boot-starter-data-rest'
//    implementation 'org.springframework.boot:spring-boot-starter-web'
    implementation 'org.springframework.cloud:spring-cloud-starter-aws-messaging'

    compileOnly 'org.projectlombok:lombok'
    developmentOnly 'org.springframework.boot:spring-boot-devtools'


    // https://mvnrepository.com/artifact/com.jayway.jsonpath/json-path
    compile group: 'com.jayway.jsonpath', name: 'json-path', version: '2.4.0'


    runtimeOnly 'com.h2database:h2'
    runtimeOnly 'mysql:mysql-connector-java'

    annotationProcessor 'org.projectlombok:lombok'
    testImplementation 'org.springframework.boot:spring-boot-starter-test'

    compile group: 'io.springfox', name: 'springfox-swagger2', version: '3.0.0-SNAPSHOT'
    compile group: 'io.springfox', name: 'springfox-swagger-ui', version: '3.0.0-SNAPSHOT'
    compile group: 'io.springfox', name: 'springfox-data-rest', version: '3.0.0-SNAPSHOT'

    testImplementation 'org.junit.jupiter:junit-jupiter-api'
    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine'
}

test {
    useJUnitPlatform()
}

task unitTest(type: Test) {
    useJUnitPlatform {
        exclude '**/*IT.class'
    }
}

task integrationTest(type: Test) {
    useJUnitPlatform {
        exclude '**/*Test.class'
    }
}

dependencyManagement {
    imports {
        mavenBom "org.springframework.cloud:spring-cloud-dependencies:${springCloudVersion}"
    }
}
```

**Comment by adnanel** [comment]
The errors are caused by version changes in the springfox dependencies (probably #c86fb6c778bbb94215db579e2838baf3340caa4a where spring boot was upgraded from 2.1 to 2.2 and spring from 5.1 to 5.2).
All our builds are faulty today, and the worst part is that the errors are runtime (at least at startup though)

To be honest I'm puzzled how such a big change didn't warrant a version increment so that we could fall back to the working snapshot.

**Comment by JKazimirski** [comment]
I suppose it is good place to attach, the reason is probably the same...

`An attempt was made to call a method that does not exist. The attempt was made from the following location:
    springfox.documentation.spring.web.plugins.DocumentationPluginsManager.createContextBuilder(DocumentationPluginsManager.java:154)

The following method did not exist:

    org.springframework.plugin.core.PluginRegistry.getPluginOrDefaultFor(Ljava/lang/Object;Lorg/springframework/plugin/core/Plugin;)Lorg/springframework/plugin/core/Plugin;`

**Comment by BiLuoHen** [comment]
if you use maven, a temporary fix:
![image](https://user-images.githubusercontent.com/12087354/71946092-c8994500-3203-11ea-951e-e87b47fc9d52.png)

**Comment by ethandeltadoritos** [comment]
Another way to temporarily make your app working is to specify the snapshot version to the latest one in 2019: 
To modify from: 
```xml
<dependency>
<groupId>io.springfox</groupId>
<artifactId>springfox-swagger2</artifactId>
<version>3.0.0-SNAPSHOT</version>
</dependency>
<dependency>
<groupId>io.springfox</groupId>
<artifactId>springfox-data-rest</artifactId>
<version>3.0.0-SNAPSHOT</version>
</dependency>
```
to
```xml
<dependency>
<groupId>io.springfox</groupId>
<artifactId>springfox-swagger2</artifactId>
<version>3.0.0-20190808.104142-39</version>
</dependency>
<dependency>
<groupId>io.springfox</groupId>
<artifactId>springfox-data-rest</artifactId>
<version>3.0.0-20190808.104142-39</version>
</dependency>
<dependency>
<groupId>io.springfox</groupId>
<artifactId>springfox-spring-web</artifactId>
<version>3.0.0-20190808.104142-38</version>
</dependency>
<dependency>
<groupId>io.springfox</groupId>
<artifactId>springfox-schema</artifactId>
<version>3.0.0-20190808.104142-39</version>
</dependency>
```

**Comment by SteamWind** [comment]
Thank you for the quick fix @ethandeltadoritos, there is a small error in your config:
```
<groupId>io.springfox</groupId>
<artifactId>springfox-spring-web</artifactId>
<version>3.0.0-20190808.104142-38</version> <--- should be 3.0.0-20190808.104142-39
```

**Comment by ccubukcu** [comment]
I had the same problem and it seems the issue stems from the version of the spring hateoas library. Spring boot 2.1.x uses hateoas 0.25.2 but the springfox version in question requires 1.0.2. I changed spring boot and data to `2.2.1.RELEASE` and spring to `5.2.1.RELEASE`. I don't have the spring-plugin dependency explicitly defined but 2.0.0 comes from springfox 3.0.0. Hope it helps.

**Comment by SteamWind** [comment]
Nope this should not be closed.

**Comment by ashishconex** [comment]
Issue still present swagger 3.0.0 does not work with spring data rest 2.3.3

**Comment by dilipkrish** [comment]
@SteamWind Is this not working with the released version? Im inclined to close this issue, unless you know for sure something is broken.

**Comment by SteamWind** [comment]
Hello, we will try it with the last release 3.0.0 next week.
At the time of my comment, this wasn't yet released.
We will keep you up to date.

**Comment by eduardo1velazquez** [comment]
If you are using an spring boot based project use version 2.2.2.RELEASE and add this as a dependency

```
<dependency>
    <groupId>io.springfox</groupId>
    <artifactId>springfox-boot-starter</artifactId>
    <version>3.0.0</version>
</dependency>
```

It worked for me

**Comment by Moufblu** [comment]
We tried, but could not manage to make it work in the time span we had. We found another solution that works with springdoc-openapi.

The changes we made are as follows:
Spring boot from 2.0.4 to 2.3.4
Springfox from 2.9.2 to none
Springdoc 1.4.8

Sorry we could not add any help.

Our project also uses spring data rest and we had some issues with all of the dependencies that had to change at the same time.
